README.md
# azure-nsg-ban-ips
Block attacker IPs for some days, perfect for Azure Kubernetes (AKS).a

This is much better than banning IPs e.g. in NGINX, 
because attackers don't reach any services at all.

Node.js: [npm package](https://www.npmjs.com/package/azure-nsg-ban-ips)

# Usage: Code

```javascript
const nsg = require( 'azure-nsg-ban-ips' )

nsg.login( SERVICE_PRICIPAL_ID, SECRET_KEY, AAD_ID, SUBSCRIPTION_ID, RESSOURCE_GROUP, NSG_NAME )

// clean up and keep ban rules 4 days
await nsg.cleanupOldBlacklists( 4 )

// add IPs to todays blacklist
let todaysBlacklist = await nsg.addIpAddrArrToBlacklist( ['1.2.3.4','6.6.6.6'] )
console.log( 'Todays IPs', todaysBlacklist )
```

Details: see [example.js](example.js) 

# Usage: Container

There is a ready to use 
- Node.js App
- [Docker Container](https://hub.docker.com/r/mahade70/aks-nsg-ingress-ban-ip)
- Kubernetes Pod (using the Docker container)
here: https://github.com/ma-ha/aks-nsg-ingress-ban-ip

This reads the NginX Ingress Controller logs, identify attacks and bans the IP addresses in the NSG.

CI code is provided for easy end-to-end set up of EventHub, LogAnalytics log feed-out etc.

# API

Init and get credentials for all following operations first:
```javascript
login( SERVICE_PRICIPAL_ID, SECRET_KEY, AAD_ID, SUBSCRIPTION_ID, RESSOURCE_GROUP, NSG_NAME )
```

Clean up NSG rules. Optional parameter: keep ban rules X days, default is 2, max is 13:
```javascript
cleanupOldBlacklists()
cleanupOldBlacklists( 4 )
```

Add IP array to todays NSG rule. Returns an array of banned IPs in todays NSG rule.
```javascript
addIpAddrArrToBlacklist( IP_ADDRESS_ARRAY )
```

Read and return all secuirity rules in the NSG:
```javascript
getNsgRules()
```
Set the base prio number for blacklist rules, e.g.
```javascript

setNsgBasePrioConfig( 1234 )
```

# Behind the Scenes

This package creates NSG security rules, for example for the NSG which comes with your AKS.

About the rules:
- one rule is created per day
- rule names are `blacklistYYYYMMDD`, for example `blacklist20201108`
- rules use prio 1000...1365 (1000 + day in year)
- rule description field contains `created by blacklist job`

The description is also used as a filter in the `cleanupOldBlacklists()` 
to identify old blacklists to delete.

If you need to set up whitelist rules, you should use a prio with e.g. 200.

CHANGELOG.md
# 0.3.1 / 0.3.2
- fix bug in `cleanupOldBlacklists()`: match names 

# 0.3.0
- new API method: `setNsgBasePrioConfig( newBasePrio )`
- fix exception handling

example.js
//const nsg = require( 'azure-nsg-ban-ips' )
const nsg = require( './azure-nsg-ban-ips' )

// change this:
const aadId   = 'YOUR AAD ID (aka tenant id)'
const spId    = 'YOUR AZURE SERVICE PRICIPAL ID'
const spKey   = 'YOUR  AZURE SERVICE PRICIPAL SECRET KEY'
const subId   = 'YOUR AZURE SUBSCRIPTION ID'
const rgName  = 'RESOURCE GROUP WHERE NSG LIVES'
const nsgName = 'YOUR NETWORK SECURITY GROUP (NSG) NAME'

run()

async function run() {
  try {
    // get credentials for all following operations first:
    await nsg.login( spId, spKey, aadId,  subId, rgName, nsgName  )

    // remove all ban rules older than 4 days
    await nsg.cleanupOldBlacklists( 4 )

    // add IPs to todays blacklist
    let todaysBlacklist = await nsg.addIpAddrArrToBlacklist( ['1.2.3.4','6.6.6.6'] )
    console.log( 'Todays IPs', todaysBlacklist )

    // get NSG rules
    let nsgRules = await nsg.getNsgRules()

    // print out ban rules
    for ( let secRule of nsgRules) {
      if ( secRule.description == 'created by blacklist job' ) {
        console.log( 'NSG ban IP rule', secRule.name, secRule.sourceAddressPrefixes )
      }
    }

  } catch ( exc ) {
    console.error( exc )
  }
}

azure-nsg-ban-ips.js
const msRestNodeAuth = require( 'ms-rest-azure' )
const NetworkManagementClient = require('azure-arm-network')
// const msRest = require( "@azure/ms-rest-azure-js" )
// const msRestNodeAuth = require( "@azure/ms-rest-nodeauth" )
// const { NetworkManagementClient } = require( "@azure/arm-network" )

var NSG_PRIO_BASE = 1000

module.exports = {
  login,
  addIpAddrArrToBlacklist,
  cleanupOldBlacklists,
  getNsgRules,
  creNSGrules,
  setNsgBasePrioConfig
}

let rg      = null
let nsg     = null
let client  = null

// ============================================================================

async function login( spId, spKey, aadId, subId, rgName, nsgName  ) {
  return new Promise( async ( resolve, reject ) => {
    try {

      // let creds = await msRestNodeAuth.loginWithServicePrincipalSecretWithAuthResponse( spId, spKey, aadId )
      // client = new NetworkManagementClient( creds, subId ) //  <<<< fails -- why??

      // msRestNodeAuth.loginWithServicePrincipalSecretWithAuthResponse( spId, spKey, aadId )
      // .then( (authres) => {
      //   // console.dir( authres, { depth: null })
      //   client = new NetworkManagementClient( authres, subId )
      // }).catch( ( err ) => {
      //   console.log( err )
      // });

      msRestNodeAuth.loginWithServicePrincipalSecret(
        spId, spKey, aadId,
        ( err, creds ) => {
          if ( err ) { return reject( err ) }
          client = new NetworkManagementClient( creds, subId )
          sub  = subId
          rg   = rgName
          nsg  = nsgName
          resolve( client )
        }
      )

    } catch ( exc ) { reject( exc ) }
  })
}

// ============================================================================

// see https://github.com/Azure/azure-sdk-for-node/blob/9ff038233ad7fd25e32b8560f899c3ce0aabc8c8/lib/services/networkManagement2/lib/operations/securityRules.js  

async function getNsgRules( ) {
  return new Promise( async ( resolve, reject ) => {
    try {
      // let client = new NetworkManagementClient( cred, sub )
      let rules = await client.securityRules.list( rg, nsg )
      resolve( rules )
        // , ( err, result, request, response ) => {
        // if ( err ) { reject( err ) }
        // resolve( result )  
      // })
    } catch ( exc ) { reject( exc ) }
  })
}

// ============================================================================

async function addIpAddrArrToBlacklist( blacklistIpArr ) {
  return new Promise( async ( resolve, reject ) => {
    try {
      let now = new Date()
      // assemble generic NSG rule name
      let secRuleName = 'blacklist' + now.getUTCFullYear() + (now.getUTCMonth() + 1) +
        ( now.getUTCDate() < 10 ? '0' : '' ) + now.getUTCDate()
      // prio must be uniqe
      let secRulePrio = NSG_PRIO_BASE + now.getUTCDay()

      // cre list of IPs to block
      let otherBlacklistIPs = await getBlockedIPs( secRuleName )
      let blacklistIPs = uniqueMerge( blacklistIpArr, otherBlacklistIPs )

      // update NSG blacklists
      if ( blacklistIPs.length > 0 ) {
        await creNSGrules( secRuleName, secRulePrio, blacklistIPs )
      }
      resolve( blacklistIPs )
    } catch ( exc ) { reject( exc ) }
  })
}

function setNsgBasePrioConfig( newBasePrio ) {
  NSG_PRIO_BASE = newBasePrio
}

// ============================================================================

// promise enabled wrapper for securityRules.createOrUpdate
async function creNSGrules( secyRuleName, secRulePrio, blacklistIpArr ) {
  return new Promise( ( resolve, reject ) => {
    // let client = new NetworkManagementClient( cred, sub )
    let secRuleParams = {
      description: 'created by blacklist job',
      priority: secRulePrio,
      protocol: '*',
      sourcePortRange: '*',
      destinationPortRange: '*',
      sourceAddressPrefix: '',
      destinationAddressPrefix: '*',
      sourceAddressPrefixes: blacklistIpArr,
      // sourceApplicationSecurityGroups:[],
      // destinationAddressPrefixes:[],
      // destinationApplicationSecurityGroups:[],
      // sourcePortRanges:[],
      // destinationPortRanges:[],
      access:'Deny',
      direction: 'Inbound',
    }
    client.securityRules.createOrUpdate(
      rg, nsg, secyRuleName, secRuleParams, 
      ( err, result, request, response ) => {
        if (err) { reject( err ) }
        //console.log( result )
        resolve( result )
      }
    )
  })
}


// ============================================================================

/* Clean up blacklists (no parma: clean up oder than 2d), max 14d  */
async function cleanupOldBlacklists( retentionDays = 2 ) {
  if ( retentionDays > 13 ) { retentionDays = 13 }
  return new Promise( async ( resolve, reject ) => {
    try {
      let fwRules = await getNsgRules( )
      let now = new Date()

      let keepRules = []
      for ( let day = 0; day < retentionDays; day++ ){
        let keepDay = new Date( Date.now() - day*24*60*60*1000 )
        let secRuleName = 'blacklist' + keepDay.getUTCFullYear() + (keepDay.getUTCMonth() + 1) +
          ( keepDay.getUTCDate() < 10 ? '0' : '' ) + keepDay.getUTCDate()
        keepRules.push( secRuleName )
      }

      // console.log( 'keep', keepRules )
      for ( let rule of fwRules ) {
        if ( rule.description === 'created by blacklist job' ) {
          if ( keepRules.indexOf( rule.name ) == -1 ) {
            try {
              // console.log( 'delete', rule.name )
              await deleteNsgRule( rule.name )
            } catch ( exc ) {
              // console.error( exc )
            }
          } //else { console.log( 'keep', rule.name ) }
        }
      }
      resolve()
    } catch ( exc ) { reject( exc ) }
  })
}


// ============================================================================

async function deleteNsgRule( secRuleName ) {
  // console.log( 'deleteNsgRule', secRuleName )
  return new Promise( ( resolve, reject ) => {
    // let client = new NetworkManagementClient( cred, sub )
    client.securityRules.deleteMethod(
      rg, nsg, secRuleName,
      ( err, result, request, response ) => {
        if (err) { reject( err ) }
        //console.log( result )
        resolve( result )
      }
    )
  })
}

// ============================================================================
// helper
function uniqueMerge( arr1, arr2 ) {
  let resultArr = arr1
  for ( let anElement of arr2 ) {
    if ( resultArr.indexOf( anElement) == -1 ) {
      resultArr.push( anElement )
    }
  }
  return resultArr
}

// ============================================================================

function getBlockedIPs( secyRuleName ) {
  return new Promise( async ( resolve, reject ) => {
    try {
      let fwRules = await getNsgRules( )
      for ( let rule of fwRules ) {
        if ( rule.name === secyRuleName ) {
          resolve( rule.sourceAddressPrefixes )
        }
      }
      resolve( [] )
    } catch ( exc)  { reject( exc ) }
  })
}